bolt-design-system/bolt
# Twig Renderer creation & initial Template rendering Jest test

## Jira

BDS-585

## Summary

Adds `@bolt/twig-renderer` which exports a `render()` and `renderString()` function that can be passed a template and data then will return an HTML string. Introduces ability to create individual template rendering snapshot testing.

## Details

See `bolt-button/__tests__/button.js` for how tests are created. In a nutshell, it's this:

```js
const { render } = require('@bolt/twig-renderer');
const { readYamlFileSync } = require('@bolt/build-tools/utils/yaml');
const { join } = require('path');
const schema = readYamlFileSync(join(__dirname, '../button.schema.yml'));
const { tag } = schema.properties;

describe('button', async () => {
  test('basic button', async () => {
    const results = await render('@bolt-components-button/button.twig', {
      text: 'Hello World',
    });
    // console.log({ results });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  tag.enum.forEach(async tagChoice => {
    test(`button tag: ${tagChoice}`, async () => {
      const results = await render('@bolt-components-button/button.twig', {
        text: 'Hello World',
        tag: tagChoice,
      });
      // console.log({ results });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });
});
```

## How to test

- A `bolt build` in `apps/bolt-site` must be ran first as this creates all the files in the data directory that the Bolt Core Twig Extension expects to find to fill `bolt.data`. You can see that the `.boltrc.js` at the top of the repo points to that same directory for reading files.
- Run `yarn test:js` (which just runs `jest`) to fire off tests. A couple commands that may be of interest:
  - `jest path/to/dir/` - will just run tests found in that dir; so we could make a `test:components` script with `jest packages/components` if we'd like. Or one could run `yarn test:js packages/components/bolt-button` to just run that single test.
  - `jest -u` will update any out of date snapshots with latest examples.
